Implements part of #17712 : Acceptance tests for CUJs in the preview tab of the exploration creator.

Akhilesh-max opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #17712 .
2. This PR does the following:  Adds the acceptance test for the following CUJs in exploration editor section.
 
  (serial numbers same as testing spreadsheet row number for each user type)
  
   6) User can preview the exploration, and can start the restart from the beginning


## Essential Checklist

- [X] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [X] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [X] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [X] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof that changes are correct



## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

Akhilesh-max committed:
exploration editor acceptance test

Akhilesh-max committed:
fixed linting failures

Akhilesh-max committed:
linting error fixed

Akhilesh-max committed:
linting errors fixed

Akhilesh-max committed:
linter

Akhilesh-max committed:
fixex

Akhilesh-max committed:
fixing

Akhilesh-max committed:
fixes

Akhilesh-max committed:
lint

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
merge

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
correction

Akhilesh-max committed:
correction

Akhilesh-max committed:
correction

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
correction

Akhilesh-max committed:
correction

Akhilesh-max committed:
correction

Akhilesh-max committed:
correction

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
flakes fixed

Akhilesh-max committed:
fixed flakes

Akhilesh-max commented:
@seanlip @gp201 @DubeySandeep PTAL

Akhilesh-max commented:
I’ve thoroughly  tested the code in both headless and non-headless modes, and it’s running well in both. I’d appreciate if you could conduct an initial review, at least for the top-level tests. 
Thanks!

seanlip commented:
@Akhilesh-max It fails in CI. Please fix before requesting a review.

Akhilesh-max committed:
fixed flakes

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
updated

Akhilesh-max committed:
lint

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
lint

Akhilesh-max committed:
parameterized the functions

Akhilesh-max committed:
lint

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint

Akhilesh-max committed:
fixes

Akhilesh-max committed:
Merge branch 'oppia:develop' into oppia-10

Akhilesh-max committed:
fixes

Akhilesh-max committed:
lints

Akhilesh-max committed:
lint fixes

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint fix

Akhilesh-max commented:
@seanlip @gp201 @DubeySandeep Both test suites are passing the CI checks. I’ve reviewed similar PRs and made efforts to ensure I have done things correctly. I would greatly appreciate your initial reviews. 
Thank you.

seanlip commented:
@Akhilesh-max If you want your PR to be reviewed, please assign the reviewers in the Assignees field or using the method provided in the PR description, and make sure the PR is not in "draft" state. Thanks.

Akhilesh-max commented:
@seanlip @DubeySandeep @gp201 @Nik-09 PTAL

seanlip commented on core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/user-factory.js:
```diff
@@ -166,6 +168,20 @@ let createNewQuestionAdmin = async function(username, email) {
+  await explorationAdmin.openBrowser();
+  await explorationAdmin.signUpNewUser(username, email);
+  activeUsers.push(explorationAdmin);
+  return explorationAdmin;
```
We don't have a role called "exploration admin" in the codebase. Maybe you mean "exploration creator" or "exploration editor"?

Akhilesh-max replied:
Made changes as per suggestion.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.js:
```diff
@@ -0,0 +1,59 @@
+  '../../puppeteer-testing-utilities/test-constants.js');
+
+const DEFAULT_SPEC_TIMEOUT = testConstants.DEFAULT_SPEC_TIMEOUT;
+const continueInteractionSelector = '.e2e-test-interaction-tile-Continue';
```
The top-level tests should not be referring to any CSS selectors.

Akhilesh-max replied:
Made changes as per suggestion.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.js:
```diff
@@ -0,0 +1,59 @@
+      await explorationAdmin.navigateToCreatorDashboard();
+      await explorationAdmin.createExploration(
+        'Exploration begins', continueInteractionSelector);
+      await explorationAdmin.LoadExplorationWithQuestions(2);
```
Don't know what "load exploration with questions" means. See the review notes on @StephenYu2018's currently-open acceptance test PR for some notes on how to structure this (he does something similar with practice questions).

Akhilesh-max replied:
I had to create a question in the exploration. Updated to use descriptive naming.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.js:
```diff
@@ -0,0 +1,59 @@
+
+      await explorationAdmin.expectTheExplorationToComplete(
+        explorationCompletedMessage);
+      await explorationAdmin.expectTheExplorationToRestart(
```
Does the user need to do something here for the exploration to restart?

Akhilesh-max replied:
Yup, It does involves a click to restart. Taken care of this.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/view-compare-and-revert-explorations.spec.js:
```diff
@@ -0,0 +1,64 @@
+      await explorationAdmin.navigateToCreatorDashboard();
+      await explorationAdmin.createExploration(
+        'Test-revision', endInteractionSelector);
+      await explorationAdmin.createMultipleRevisionsOfTheSameExploration(
```
This sounds like you're making multiple copies. 

I also think each save should be a separate action since that is how an end user would describe it -- "make this change, then save".

Akhilesh-max replied:
Sure, separated into multiple steps to handle this.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/view-compare-and-revert-explorations.spec.js:
```diff
@@ -0,0 +1,64 @@
+        'Test-revision', 13);
+
+      await explorationAdmin.navigateToHistoryTab();
+      await explorationAdmin.expectRevisionsToHave(
```
This is a bit vague. I think you'll want to do something like: expect there to be N revisions, expect first revision to have these details, expect second revision to have these details, then play with filtering/pagination.

Akhilesh-max replied:
@seanlip @gp201, if we have multiple revisions, would it be adequate to check if the latest revision has all the necessary properties such as version number, notes, username, and date? If not, would the following be intuitive to the end user: expectAllRevisionsToHaveVersionNoNotesUsernameDate()? Or should I break it down for each revision that I have?

seanlip replied:
OK to check just a couple of revisions, as per my first comment. I think I'd suggest checking 2 rather than 1 just to make sure multiple of them display.

The method name you propose is not really understandable, but that's probably moot given what I mentioned in the previous paragraph.

Akhilesh-max replied:
sure, thanks.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/view-compare-and-revert-explorations.spec.js:
```diff
@@ -0,0 +1,64 @@
+      await explorationAdmin.filterRevisionsByUsername('explorationAdm');
+      await explorationAdmin.ExpectPaginatorToChangeItemsPerPage(15);
+
+      await explorationAdmin.compareDifferentRevisions();
```
These functions should correspond to user actions and when the user is doing the action they are going to be picking specific revisions. So the revision numbers should be entered as arguments to the function.

In general if the user's actions can be parameterized then the functions should reflect that.

Akhilesh-max replied:
sure, done. Altered it to accept it the revision number.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/view-compare-and-revert-explorations.spec.js:
```diff
@@ -0,0 +1,64 @@
+      await explorationAdmin.ExpectPaginatorToChangeItemsPerPage(15);
+
+      await explorationAdmin.compareDifferentRevisions();
+      await explorationAdmin.expectCompareToDisplayMetadataChanges();
```
All "expects" should be observations. What does the user see? Show that in your expectation -- e.g. expectMetadataChangesToInclude(...)

Akhilesh-max replied:
done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor.utils.js:
```diff
@@ -0,0 +1,517 @@
+  * @param {string} text - the text of the Exploration.
+  * @param {number} numOfVersions - number of versions to created.
+  */
+  async createMultipleRevisionsOfTheSameExploration(text, numOfVersion) {
```
Here's how to think about functions in this file -- have each top-level function here reflect common actions of the end user, E.g. saving an exploration, adding a question, etc. Each action should be describable in a single short phrase and be something an end user would naturally do.

If an end user would not naturally do "make 10 revisions of this exploration" then don't have that function here. That corresponds to the end user *choosing* to make 10 separate edits -- those choices should be reflected in the "user" layer of the tests. But if the user's path for an action is constrained (e.g. they want to add a piece of feedback and that path involves clicking X, typing Y, clicking Z) -- then it's fine to have that entire action as a function in this file.

Hope this gives a helpful heuristic about how to structure the test between the different layers.

Akhilesh-max replied:
@seanlip Oh, it did. Thank you for the guidance! I’m learning.

However, I’m concerned that representing each edit separately (which would be more natural to the user) might increase the number of tests in the top-level file. Would we be ok with that? 
Moreover, I created multiple revisions primarily to verify the paginator’s functionality. Would it be an adequate test to ensure that the next page button stays disabled when changing the items per page on the paginator? This method would allow me to test the paginator without the need to create a large number of items. Does this seem reasonable?

seanlip replied:
Do you mean "increase the number of statements"? Yes I'm OK with that; it would help if there was a comment on top of that "block" of statements to explain what the user is doing.

I don't understand what you are trying to do in your second paragraph. Maybe specify clearly what behaviour you are trying to test, exactly.

Akhilesh-max replied:
@seanlip  @gp201 Suppose I need to test a paginator. Would it be reasonable to set the paginator to display 15 revisions per page? If there are more than 15 items, the ‘next page’ (>) button should be enabled; otherwise, it should not. So, if I don’t have 15 revisions and I set the paginator to display 15 revisions per page, would checking whether the ‘next page’ (>) button is enabled or not be a good test for the paginator? if so, would that be a sufficient check?
Thanks!

<img width="384" alt="Screenshot 2024-03-12 at 3 33 36 PM" src="https://github.com/oppia/oppia/assets/124369508/38570f3c-0fd6-4544-b57e-835de09751f8">

seanlip replied:
I don't know. What behaviour are you testing -- what is/are the specific behaviour(s) for the end user that you want to preserve through these tests? State them like this: "As an exploration creator, if I am looking at the history tab for an exploration that has XXX revisions, I should be able to ABC."

Akhilesh-max replied:
@seanlip,
Let me rephrase my concern.

As an exploration creator, if I am looking at the history tab for an exploration
that has more than 10 revisions, I should be able to navigate to the next page of revisions when the paginator is set to display 10 revisions per page.

Here’s the behavior I want to verify:

1) If there are more than 10 revisions, the ‘next page’ (>) button should be enabled, allowing the user to navigate to the next page of revisions.
2) If there are fewer than 10 revisions, the ‘next page’ (>) button should be disabled as there are no additional pages to navigate to.

My question is: Would checking the state of the ‘next page’ (>) button (whether it’s enabled or disabled) be a sufficient test for this paginator functionality?
 Or should I include additional checks? If so, may you please suggest?

seanlip replied:
I would say that your check is reasonable, but I would also add clicking on the button in the "enabled" case and verifying that the right set of results show up (because that is what a user would do).

Akhilesh-max replied:
sure, will take care.

seanlip reviewed (changes requested):
Thanks @Akhilesh-max -- took a quick pass, left some high-level comments.

Akhilesh-max commented:
@seanlip Since we’ve migrated to TypeScript for these tests, should I write this test in TypeScript only?

seanlip commented:
Yes, please write in typescript, thanks.

DubeySandeep reviewed (approved):
LGT for the changes in `scripts/common.py`

Akhilesh-max committed:
conflicts resolved

Akhilesh-max committed:
resolved conflicts

Akhilesh-max committed:
updated branch

Akhilesh-max committed:
migrated to ts

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
type fix

Akhilesh-max committed:
updated branch

Akhilesh-max committed:
lint fix

Akhilesh-max committed:
update

Akhilesh-max committed:
refactoring

Akhilesh-max committed:
pulled changes

Akhilesh-max committed:
changes

Akhilesh-max committed:
updating

Akhilesh-max committed:
lint

Akhilesh-max committed:
correction

Akhilesh-max committed:
updated

Akhilesh-max committed:
lint

Akhilesh-max committed:
lint

Nik-09 reviewed (approved):
Thanks @Akhilesh-max LGTM

Akhilesh-max commented:
@seanlip As I’m still refining the utility files, I foresee that any modifications (based on reviews) in the top-level/spec files might necessitate further tweaks to the utility file. This is because I’ll need to adapt them based on the updated top-level file, leading to additional work. If this sounds reasonable to you, may you please only review the top-level file again for now?

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+    'should be able to load, complete and restart an exploration preview',
+    async function () {
+      await explorationEditor.navigateToCreatorDashboard();
+      await explorationEditor.createExploration('Exploration begins');
```
This is an odd name for an exploration. Pick something that a creator will be more likely to use.

Akhilesh-max replied:
Corrected.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      await explorationEditor.navigateToCreatorDashboard();
+      await explorationEditor.createExploration('Exploration begins');
+      await explorationEditor.addAnInteractionToTheExploration(
+        ' Continue Button '
```
This should be some named constant picked from an enum somewhere, like INTERACTION_TYPES.CONTINUE

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+    async function () {
+      await explorationEditor.navigateToCreatorDashboard();
+      await explorationEditor.createExploration('Exploration begins');
+      await explorationEditor.addAnInteractionToTheExploration(
```
explorationEditor.addInteraction(INTERACTION_TYPES.CONTINUE)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      );
+
+      // The following block of code adds a new card to the exploration that contains a question.
+      await explorationEditor.addANewCardToTheExploration('Question Card');
```
I don't think this is actually what the exploration editor does. What action do they take here -- adding an answer group or modifying the existing one?

Akhilesh-max replied:
There are some steps in between, taken care of.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+
+      // The following block of code adds a new card to the exploration that contains a question.
+      await explorationEditor.addANewCardToTheExploration('Question Card');
+      await explorationEditor.goToTheCard('Question Card ');
```
explorationEditor.navigateToCard(...)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      // The following block of code adds a new card to the exploration that contains a question.
+      await explorationEditor.addANewCardToTheExploration('Question Card');
+      await explorationEditor.goToTheCard('Question Card ');
+      await explorationEditor.addContentToTheCard(
```
explorationEditor.updateCardContent(....)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      await explorationEditor.addContentToTheCard(
+        'mention a negative number greater than -100.'
+      );
+      await explorationEditor.addAnInteractionToTheCard(' Number Input ');
```
explorationEditor.addInteraction(INTERACTION_TYPES.NUMERIC_INPUT) (or whatever the actual name of the interaction is, as shown on the screen)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+        'mention a negative number greater than -100.'
+      );
+      await explorationEditor.addAnInteractionToTheCard(' Number Input ');
+      await explorationEditor.addResponsesToTheInteraction('-45');
```
There's only one response here, so "add responses" is misleading.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      await explorationEditor.addContentToTheCard('Exploration ends here');
+      await explorationEditor.addAnInteractionToTheCard(' End Exploration ');
+
+      await explorationEditor.goToTheIntroductionCard(0);
```
explorationEditor.navigateToCard(/* put the state name here */)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      await explorationEditor.saveExplorationDraft();
+
+      await explorationEditor.navigateToPreviewTab();
+      await explorationEditor.expectTheExplorationToStartFromIntroductionCard(
```
This isn't what the user sees. What exactly is the user looking at the screen and trying to validate?

Akhilesh-max replied:
Changed it to expect the current card content on the screen.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+        'Exploration begins'
+      );
+
+      await explorationEditor.completeTheExplorationInPreviewTab();
```
This is likely to depend on the actual structure of the exploration, you might need to break this up into several "answer the question" or "click continue" actions.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      );
+
+      await explorationEditor.completeTheExplorationInPreviewTab();
+      await explorationEditor.expectTheExplorationToComplete();
```
Not clear what's being checked for here. Try to explain what precisely the user sees / is looking for on the screen.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+
+      await explorationEditor.completeTheExplorationInPreviewTab();
+      await explorationEditor.expectTheExplorationToComplete();
+      await explorationEditor.restartTheExploration();
```
restartExploration()

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,79 @@
+      await explorationEditor.completeTheExplorationInPreviewTab();
+      await explorationEditor.expectTheExplorationToComplete();
+      await explorationEditor.restartTheExploration();
+      await explorationEditor.expectTheExplorationToRestart(
```
Not sure what's being checked for here, again try to explain what the user is seeing / looking for on the screen.

Akhilesh-max replied:
Changed it to expect the Introduction card content, upon restart.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,663 @@
+   * Function to navigate to the settings tab in the Exploration Editor.
+   */
+  async navigateToSettingsTab(): Promise<void> {
+    await this.page.waitForTimeout(200);
```
We won't accept PRs that use waitForTimeout(). Always wait for a specific event to occur, similar to the waitFor in the webdriverio e2e tests.

Akhilesh-max replied:
Sure, taken care.

seanlip reviewed (changes requested):
Thanks @Akhilesh-max -- I did a top-level review.

Akhilesh-max committed:
updates

Akhilesh-max committed:
updated

Akhilesh-max committed:
update

Akhilesh-max committed:
correct

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
fixig flakes

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

Akhilesh-max committed:
fixes

Akhilesh-max committed:
fixes

Akhilesh-max committed:
fixes

Akhilesh-max committed:
fixes

Akhilesh-max commented:
Hi @seanlip, I’ve decided to address one CUJ in this PR and will cover the other in a separate PR. I’ve updated the PR description accordingly. 
I’ve also addressed all your previous comments (you had reviewed only one CUJ last time, which I’m covering in this PR). Also, both the mobile and desktop tests are passing the CI.

Please have a look when you get a chance. Thanks!

Akhilesh-max commented:
@seanlip PTAL

seanlip commented on core/templates/pages/exploration-editor-page/editor-tab/graph-directives/exploration-graph.component.html:
```diff
@@ -43,7 +43,7 @@
     </mat-card>
   </div>
-  <div class="oppia-exp-graph-modal-icon-container" (click)="openStateGraphModal()">
+  <div class="oppia-exp-graph-modal-icon-container e2e-test-oppia-mobile-graph-resize-button" (click)="openStateGraphModal()">
```
In other class names we don't use 'oppia' so maybe we should omit it here too.

Akhilesh-max replied:
dfd

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/user-factory.ts:
```diff
@@ -138,10 +142,15 @@ export class UserFactory {
+  ): Promise<
+    LoggedInUser & ExplorationEditor & MultipleRoleIntersection<TRoles>
+  > {
+    let user = UserFactory.composeUserWithRoles(
```
I don't understand why you need a nested composeUserWithRoles() here. /cc @jnvtnguyen

Akhilesh-max replied:
Actually, true, the nested composeUserWithRoles here is unnecessary as we can assign multiple roles in a single call. I've updates it to call composeUserWithRoles once with both LoggedInUserFactory and ExplorationEditorFactory passed in the roles array.

Thanks.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+import testConstants from '../../puppeteer-testing-utilities/test-constants';
+import {ExplorationEditor} from '../../user-utilities/exploration-editor-utils';
+
+const DEFAULT_SPEC_TIMEOUT: number = testConstants.DEFAULT_SPEC_TIMEOUT;
```
Can you update both this and the original constant to have a unit in the name (e.g. DEFAULT_SPEC_TIMEOUT_SECS or DEFAULT_SPEC_TIMEOUT_MSECS)

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+
+const DEFAULT_SPEC_TIMEOUT: number = testConstants.DEFAULT_SPEC_TIMEOUT;
+const INTRODUCTION_CARD_CONTENT: string =
+  'Test exploration to question negative numbers';
```
I don't understand what this means, try and use something realistic.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+const DEFAULT_SPEC_TIMEOUT: number = testConstants.DEFAULT_SPEC_TIMEOUT;
+const INTRODUCTION_CARD_CONTENT: string =
+  'Test exploration to question negative numbers';
+const NEW_CARD: string = '/';
```
I don't understand this constant. Firstly, something named "card" should be a card, not a string. If you mean the card name, then I don't think we have cards that are named "/". So overall I don't really understand what is going on here.

Akhilesh-max replied:
Here, the NEW_CARD constant is actually not representing any card or a card name. Instead, it represents a value used in the UI’s dropdown menu to select the option for creating a new card. This selection is made using our custom select() function, which asks for the value of the option to be selected. And, the ‘/’ value refers to the value assigned to that option in the dropdown menu.
However, since now I have merged the two functions, this is no longer needed in the top-level file. I have also explained the use of this in a comment on the function `addResponsesToTheInteraction` and `directLearnersToNewCard.`

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+  'Test exploration to question negative numbers';
+const NEW_CARD: string = '/';
+enum INTERACTION_TYPES {
+  CONTINUE_BUTTON = ' Continue Button ',
```
Why are there spaces around the texts?

Akhilesh-max replied:
Actually, I was using the exact format of the interaction names as they appear in the UI to click on them, which included spaces. This was reflected in the enum values.
However, to keep the enum clean and reusable, I've now moved this formatting directly into the function addInteraction where these enum values are used. I've also added a comment in the function to explain this adjustment.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+  beforeAll(async function () {
+    explorationEditor = await UserFactory.createNewUser(
+      'explorationEditor',
+      'exploration_creator@example.com'
```
Maybe have the email addresses roughly follow the usernames for parity, e.g. exploration_editor@example.com in this case.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+    async function () {
+      // Navigate to the creator dashboard and create a new exploration.
+      await explorationEditor.navigateToCreatorDashboard();
+      await explorationEditor.createExploration();
```
The current name is misleading. This call doesn't create an exploration, it just clicks the create exploration button.

Akhilesh-max replied:
changes to `clickCreateExplorationButton()`

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+
+      // Add a new card with a question.
+      await explorationEditor.viewOppiaResponses();
+      await explorationEditor.oppiaDirectlearnersTo(NEW_CARD);
```
Methods should start with a verb.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+      // Add a new card with a question.
+      await explorationEditor.viewOppiaResponses();
+      await explorationEditor.oppiaDirectlearnersTo(NEW_CARD);
+      await explorationEditor.nameNewCard('Test Question');
```
I think you can combine this with the previous call since conceptually they are a single action (direct learners to new card named XXX)

Akhilesh-max replied:
combined. directLearnersToNewCard()

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+        'Enter a negative number greater than -100.'
+      );
+      await explorationEditor.addInteraction(INTERACTION_TYPES.NUMERIC_INPUT);
+      await explorationEditor.addResponseToTheInteraction('-99');
```
This doesn't seem like enough info to specify a response, don't we need to include correctness status, feedback, destination as part of doing so?

Conceptually, maybe some stuff in the next block should be grouped with this.

Akhilesh-max replied:
Updated. It looks something like this now:

addResponsesToTheInteraction( interactionType: string, answer: string, feedback: string, destination: string, correctness: string ): Promise<void>

Adding answer to an interaction, depends on the interaction type, so included that as well.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+      await explorationEditor.addResponseToTheInteraction('-99');
+
+      // Add a final card with an 'End' interaction.
+      await explorationEditor.oppiaDirectlearnersTo(NEW_CARD);
```
Possibly FINAL_CARD is a better name for this, so that readers of this code can easily identify it as such.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+  }, DEFAULT_SPEC_TIMEOUT);
+
+  it(
+    'should be able to load, complete and restart an exploration preview',
```
If most of the test is about testing the preview functionality, then should the creation of the exploration be part of the setup step instead?

Akhilesh-max replied:
Thanks for this suggestion. Moving the creation of the exploration to the setup step as it's more of a prerequisite for the test rather than the test itself.

However, I do have a concern. If we move all statement before the navigation to the preview tab into the setup or beforeAll block, and an error occurs, it will appear as a suite error. The execution will still move into the 'it' block (which will obviously fail since the setup was erroneous), leading to two error messages, of which only the latter is relevant.

Despite this, I've proceeded with the change as it does make the test more readable and logically structured. Looking for you suggestion, whether we should stick with this approach or consider a prior one.

seanlip replied:
Thanks for catching that. Can you look into why the execution moves into 'it' even if beforeAll fails? IMO the test should just stop if the beforeAll fails -- maybe there is a way to configure that behaviour.

Akhilesh-max replied:
sure, will raise a issue regarding and try to look into the solution.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration.spec.ts:
```diff
@@ -0,0 +1,109 @@
+      // Continue to the next card, enter an answer, and submit it.
+      await explorationEditor.continueToNextCard();
+      await explorationEditor.enterAnswer('-40');
+      await explorationEditor.submitAnswer();
```
I think this can be combined with the previous call (`submitAnswer('-40')`).

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+  async updateCardContent(questionText: string): Promise<void> {
+    await this.clickOn(stateEditSelector);
+    await this.waitForElementToBeClickable(stateContentInputField);
+    await this.page.click(stateContentInputField, {clickCount: 3});
```
Why do you need to click this 3 times?

Akhilesh-max replied:
I had implemented this before `waitForElementToBeClickable()` function got incorporated into the `clickOn()` function. This was to ensure that the same input component did not retain previous entries. However, I didn’t verify that this too was fixed post updating the `clickOn()` function.
I have now removed this.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+    await this.clickOn(stateEditSelector);
+    await this.waitForElementToBeClickable(stateContentInputField);
+    await this.page.click(stateContentInputField, {clickCount: 3});
+    await this.page.keyboard.press('Backspace');
```
Why do you need this step?

Akhilesh-max replied:
Same reason as mentioned above. However, not relevant any more. Removed this as well

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+   */
+  async viewOppiaResponses(): Promise<void> {
+    await this.page.waitForSelector(stateResponsesSelector, {visible: true});
+    await this.page.click(stateResponsesSelector);
```
Why do we use this.page.click in some places and this.clickOn in others? Can we be consistent throughout and also with the other tests?

Akhilesh-max replied:
Done. Using only `clickOn()` across.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+  /**
+   * Function to select the card that learners will be directed to from the current card.
+   */
+  async oppiaDirectlearnersTo(card: string): Promise<void> {
```
Functions should start with a verb, and also "direct learners" is two words.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+        await this.clickOn(mobileOptionsButton);
+      }
+      await this.clickOn(mobileSaveChangesButton);
+      await this.clickOn(saveDraftButton);
```
Pull common statements like this and the following line outside the if/else. In general try to reduce the branching.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+   * @param {string} cardName - The name of the card to navigate to.
+   */
+  async navigateToCard(cardName: string): Promise<void> {
+    if (cardName === 'Introduction' && this.isViewportAtMobileWidth()) {
```
This won't do, you need to make this function more general and not assume a specific format of the exploration. It should error if the name isn't in the graph and generically navigate to the right state otherwise.

Akhilesh-max replied:
Yes, the previous implementation was kinda vague and limited. I have tried to generalise it. The function now receives two parameters: one is the list of card names, and the other is the specific card to navigate to.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+    await this.type(stateContentInputField, 'Correct Answer, You got that!');
+    await this.clickOn(correctAnswerInTheGroupSelector);
+    await this.clickOn(addNewResponseButton);
+    await this.page.waitForSelector(resonseModalHeaderSelector, {hidden: true});
```
"response" is misspelled.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+      await this.page.waitForNavigation();
+    } else {
+      await this.clickOn(previewTabButton);
+      await this.page.waitForNavigation();
```
Same note about moving common stuff outside the if/else

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+    );
+    if (introMessage === text) {
+      showMessage(
+        'Preview is on the Introduction card and is loading correctly.'
```
It's unclear from the name of this function that (a) it is being called in the context of the preview card, and (b) the text specifically refers to the Introduction card. Please make this utility function more general and name it clearly to indicate the scope in which it should be used.

Akhilesh-max replied:
Addressed both the concerns and tried making the function more general.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+  }
+
+  async enterAnswer(answer: string): Promise<void> {
+    await this.type(floatFormInput, answer);
```
This assumes the answer is a float, which is not necessarily true. You might need to scan the DOM to figure out what element is on the page and handle things accordingly based on the element. It's fine to do this for a few common element types for now and note the limitations in the docstring.

Akhilesh-max replied:
Made changes as per suggestion. I'm now taking care of string, number and float. Also, mentioned the same in the docstring.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+      if (element) {
+        await this.clickOn(mobileOptionsButton);
+      }
+      await this.clickOn(previewRestartButton);
```
Same note about extracting common functionality.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,288 @@
+  async restartPreview(): Promise<void> {
+    if (this.isViewportAtMobileWidth()) {
+      const element = await this.page.$(mobileNavbarOptions);
+      if (element) {
```
Can you describe, in a comment, why this element might not exist?

Akhilesh-max replied:
when the mobile navigation bar is expanded, it actually, overlaps the restart button making it unclickable. So, checking for its presence and collapsing it. Mentioned the same in a comment.

Can it make a issue?

seanlip replied:
Yes, please file an issue, thanks. (In general, please file an issue any time you see something is not working as expected -- no need to ask beforehand.)

Akhilesh-max replied:
Done.

seanlip replied:
@Akhilesh-max Please link to the issue you filed. (It's a good idea to do this in general when this sort of thing comes up in review comments.)

Akhilesh-max replied:
Done

seanlip commented on .github/workflows/e2e_lighthouse_performance_acceptance_tests.yml:
```diff
@@ -312,6 +312,7 @@ jobs:
           suite:
             - blog-admin-tests/assign-roles-to-users-and-change-tag-properties
             - blog-editor-tests/try-to-publish-a-duplicate-blog-post-and-get-blocked
+            - exploration-editor-tests/load-complete-and-restart-exploration
```
Not sure about the name -- you're mainly testing the preview functionality, right? So, perhaps, `preview-exploration` ?

Akhilesh-max replied:
Yes, the tests are primarily testing the preview functionality. However, I’m curious if the preview-exploration includes restarting of the exploration? For now, I have updated it to load-complete-and-restart-exploration-preview.
Looking for your suggestion.

seanlip replied:
I think what you have is probably fine. Thanks!

Akhilesh-max replied:
sure, thanks.

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max commented:
@seanlip PTAL

seanlip commented:
@Akhilesh-max Please reply to all comments, following the instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve)? In particular, for comments that you've addressed, please reply "Done." -- this helps keep track of which comments have been addressed and which haven't. Thanks!

Akhilesh-max commented:
@seanlip Sorry, I see, I missed one comment...

Akhilesh-max committed:
updates

Akhilesh-max commented:
@seanlip Addressed all the remaining comments. Please take a look.

Thank you.

seanlip commented:
No, there are still no replies to my comments. Please take a look at your PR in an incognito window to see how it looks to other reviewers.

Akhilesh-max commented:
@seanlip So sorry, like really sorry. I was doing something really silly because of which the comments were not visible to you. 
PTAL

Thanks.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration-preview.spec.ts:
```diff
@@ -0,0 +1,115 @@
+      '-99',
+      'Prefect!',
+      'Final Card',
+      'correct'
```
This last arg should be a boolean.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/spec/exploration-editor-tests/load-complete-and-restart-exploration-preview.spec.ts:
```diff
@@ -0,0 +1,115 @@
+      INTERACTION_TYPES.NUMERIC_INPUT,
+      '-99',
+      'Prefect!',
+      'Final Card',
```
CARD_NAME.FINAL_CARD

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,331 @@
+  async saveExplorationDraft(): Promise<void> {
+    if (this.isViewportAtMobileWidth()) {
+      const element = await this.page.$(mobileNavbarOptions);
+      if (!element) {
```
In what circumstance is there no element present? Maybe explain that in a comment.

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,331 @@
+   * @param {string} cardName - The name of the card to navigate to.
+   * @param {string[]} cardNames - An array of all card names in the exploration.
+   */
+  async navigateToCard(cardName: string, cardNames: string[]): Promise<void> {
```
Can you implement this function without passing in cardNames? (From the perspective of an end user, they should be able to look at the graph on the page and find the one corresponding to the desired card name without knowing what the other card names are, right?)

Akhilesh-max replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,331 @@
+   */
+  async expectPreviewCardContentToBe(
+    cardName: string,
+    text: string
```
expectedCardContent perhaps (put "expected" before values that communicate the expected behaviour, and match the names -- don't call one thing "content" and the other "text" if they both refer to the same thing).

Akhilesh-max replied:
Done

Akhilesh-max commented:
@seanlip  
Regarding the issue of continued execution in the ‘it’ block despite a failure in the ‘beforeAll’ block, this problem can be resolved by setting the stopOnSpecFailure flag to true in the ‘jasmine.json’ file. However, enabling this flag will prevent the execution from proceeding to any other spec for all the jasmine based test. can we do this?

<img width="578" alt="Screenshot 2024-04-16 at 3 24 31 PM" src="https://github.com/oppia/oppia/assets/124369508/18b68a6f-5d0b-4397-aac2-a90f706de03f">

seanlip commented:
@Akhilesh-max If this is per-suite then stopping on the first failure is fine. It might be nice to do this for webdriverio as well.

Yes, OK to create an issue for this and a separate PR to solve it, thanks.

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max commented:
@seanlip PTAL.

Akhilesh-max commented:
@seanlip, A recent Acceptance test [PR](https://github.com/oppia/oppia/pull/19994) has introduced utility file for exploration editor, shell I move all my utility functions into that?

Akhilesh-max committed:
updates

seanlip commented:
> @seanlip, A recent Acceptance test [PR](https://github.com/oppia/oppia/pull/19994) has introduced utility file for exploration editor, shell I move all my utility functions into that? Or, should I simple rename the utility file accordingly?

@Akhilesh-max You should probably have all "exploration creator/editor" actions in a single file. I don't have a strong opinion on whether it should be called exploration-creator-utils or exploration-editor-utils, as long as we're consistent throughout (e.g. don't call the role ExplorationCreator and name the file exploration-editor-utils, for example).

seanlip commented:
> @Akhilesh-max If this is per-suite then stopping on the first failure is fine. It might be nice to do this for webdriverio as well.
> 
> Yes, OK to create an issue for this and a separate PR to solve it, thanks.

Just a note, please link to the issue in this PR once you've created it, thanks.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -127,6 +129,9 @@ export class ExplorationEditor extends BaseUser {
     if (this.isViewportAtMobileWidth()) {
       const element = await this.page.$(mobileNavbarOptions);
+      // If the element is not present, it means the mobile navigation bar is not expanded.
+      // The option to save changes in mobile width, appears only after clicking on the mobile options button,
```
...in mobile view appears only ...

Akhilesh-max replied:
Done.

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -169,14 +183,14 @@ export class ExplorationEditor extends BaseUser {
     feedback: string,
     destination: string,
-    correctness: string
+    correctness: boolean
```
correctness --> responseIsCorrect

Akhilesh-max replied:
Done

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -169,14 +183,14 @@ export class ExplorationEditor extends BaseUser {
-   * @param {string} correctness - Whether the response is marked as correct.
+   * @param {boolean} correctness - Whether the response is marked as correct.
    */
   async addResponsesToTheInteraction(
```
Aren't you adding just one response? If so then why is the function name plural?

Akhilesh-max replied:
oh, certainly, the name shouldn't be plural. 
Changed to  addResponseToTheInteraction().

seanlip reviewed (changes requested):
Just a note, there are still unaddressed comments here, PTAL: https://github.com/oppia/oppia/pull/19828#pullrequestreview-1956404709

Akhilesh-max commented:
#19916 issue :  preview restart button not clickable when mobile navigation bar is expanded.

Akhilesh-max commented:
#20175, Issues of the continued execution of test despite initial failures.

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max commented:
I'm using 'exploration editor' everywhere for the shake of consistency. Also, on the website, we call it the same. 
<img width="473" alt="Screenshot 2024-04-18 at 2 37 14 AM" src="https://github.com/oppia/oppia/assets/124369508/5d5c0dd8-b30a-4da5-b0af-73508325aeef">

Akhilesh-max commented:
@seanlip PTAL

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

seanlip commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -62,17 +77,87 @@ const explorationCompletionToastMessage = '.e2e-test-lesson-completion-message';
+      await this.clickOn(mobileChangesDropdown);
+      await this.clickOn(mobilePublishButton);
+    } else {
+      await this.page.waitForSelector(
```
Can we use the utility function waitForElementToBeClickable ?

Akhilesh-max replied:
Here, I guess, It's fine without the existing one and the utility function (waitForElementToBeclickable()) as well, since the clickOn() already incorporates this waitForElementToClickable(). 
Updated accordingly. 
Thanks.

seanlip reviewed (changes requested):
@Akhilesh-max There are still old comments you haven't responded to. PTAL at the Conversation page, expand all comments, and ensure that each has a reply. Thanks.

Akhilesh-max committed:
updates

Akhilesh-max committed:
updates

Akhilesh-max commented:
@seanlip PTAL.

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

seanlip reviewed (approved):
Thanks @Akhilesh-max. LGTM!

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

Akhilesh-max committed:
Merge branch 'develop' into oppia-10

Akhilesh-max commented:
@seanlip Shell we get this merged?
Thank you.

rahat2134 commented on core/tests/puppeteer-acceptance-tests/user-utilities/exploration-editor-utils.ts:
```diff
@@ -0,0 +1,432 @@
+      case 'Number Input':
+        await this.type(floatFormInput, answer);
+        break;
+      // Add cases for other interaction types here
+      // case 'otherInteractionType':
+      //   await this.type(otherFormInput, answer);
+      //   break;
```
@Akhilesh-max I think it needs to be corrected. PTAL.
